### PR TITLE
Pull in some DoS mitigations from upstream

### LINKED ISSUE #1251: Merge upstream anti DoS patches
At the time of writing, testnet is at block 4109 but the alpha node is handing out many peers who continuously advertise a block height of 1337 and are not synced.  The result is that genuine nodes receiving these peers are unable to join the network.

```
    {
        "id" : 9,
        "addr" : "x.x.x.x:18233",
        "addrlocal" : "x.x.x.x:55684",
        "services" : "0000000000000001",
        "lastsend" : 1471152190,
        "lastrecv" : 1471152189,
        "bytessent" : 1719575,
        "bytesrecv" : 947,
        "conntime" : 1471152188,
        "timeoffset" : -1,
        "pingtime" : 0.00000000,
        "pingwait" : 31.86533400,
        "version" : 170002,
        "subver" : "/Satoshi:0.11.2/",
        "inbound" : false,
        "startingheight" : 1337,
        "banscore" : 0,
        "synced_headers" : -1,
        "synced_blocks" : -1,
        "inflight" : [
        ],
        "whitelisted" : false
    },

```

Upstream improvements should be identified and where possible merged.  For example "Connection slot exhaustion DoS mitigation" https://github.com/bitcoin/bitcoin/pull/6374

### PR DESCRIPTION
**Pull request by defuse**
Closes #1251.

**WARNING: I force pushed**

### CODE REVIEW

**Comment by defuse** [comment]
This doesn't compile.

**Comment by defuse** [comment]
The upstream PR was made against a better version of the code (e.g. comparing our code to the upstream PR's old code in the diff, there are comments there that doesn't exist in ours), but there's a couple places our code was ahead, I'm not sure why:

1.

```
 +                    // According to the internet TCP_NODELAY is not carried into accepted sockets
 +                    // on all platforms.  Set it again here just to be sure.
 +                    int set = 1;
 +#ifdef WIN32
 +                    setsockopt(hSocket, IPPROTO_TCP, TCP_NODELAY, (const char*)&set, sizeof(int));
 +#else
 +                    setsockopt(hSocket, IPPROTO_TCP, TCP_NODELAY, (void*)&set, sizeof(int));
 +#endif
 +
```

from 95a50390, trying to use some macros/functions which don't exist in our code.

2. The `-whiteconnections` switch that the upstream PR removes was already gone in our code.

That's kind of weird...

**Comment by defuse** [comment]
Oh crap, I cherry picked the wrong commit. Fixed and rebased.

**Comment by defuse** [comment]
This should be ready for review.

**Comment by bitcartel** [comment]
Is the commit message "Refactor: AcceptConnection (modified to include 95a5039)" valid, given that the changes from 95a5039 have been backed out.   Can we fix the message?

**Comment by defuse** [comment]
What I meant by that commit message was "We have https://github.com/zcash/zcash/commit/95a50390e1052a0a501eb446f87d63f58d95b7e7#diff-9a82240fe7dfe86564178691cc57f2f1R872 in our current codebase already, and it caused a conflict with the PR I'm backporting, so to fix that conflict I manually applied (part of) https://github.com/zcash/zcash/commit/95a50390e1052a0a501eb446f87d63f58d95b7e7#diff-9a82240fe7dfe86564178691cc57f2f1R872 to the area that was in conflict."

File: src/net.cpp [hunk 1]
```diff
@@ -671,6 +671,231 @@ void SocketSendData(CNode *pnode)
 
 static list<CNode*> vNodesDisconnected;
 
+class CNodeRef {
```
Comment (daira) [inline src/net.cpp]: Ugh, is this just a refcounting wrapper around a ``CNode*``? I know it's better to follow upstream exactly here, but if we needed anything like this, I'd like to think that we'd just use ``shared_ptr``.

Comment (str4d) [inline src/net.cpp]: `shared_ptr` was introduced in C++ 11, so it makes sense that the upstream PR wouldn't have included it.

**Comment by daira** [comment]
The heuristics here don't seem unreasonable, but how do we know that they'll actually fix the current DoS? They appear to have been designed under the assumption that there are a lot of legitimate nodes to connect to, whereas we are overdependent on alphatestnet.z.cash.

**Comment by daira** [comment]
ACK because I don't think this will do any harm, and any further measures we take should start from this.

**Comment by defuse** [comment]
I'm also unsure if this will fix the current DoS. I think we'll have to move to a better testnet architecture, so I left a comment in #630 asking to clarify whether we'll use that same bootstrapping stuff for the testnet.

**Comment by ebfull** [comment]
ACK, I'm glad we can bring this stuff in safely.

**Comment by daira** [comment]
@zkbot r+

**Pull request merged**